[MAINTENANCE] Consolidate different Metric Types definition Enums

alexsherstinsky opened the pull request:
### Scope
* Currently different metric types `Enum` classes are either in `ExecutionEngine` and/or scattered.  This change consolidates all of them into properly-named dedicated modules.
* Type hint fixes in `RuleBasedProfiler` components (completion from previous effort).

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1358/GREAT-1527
-
-


After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-


### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.


Thank you for submitting!

Commit:
refactor metrics types

Commit:
merge

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=6746&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=6746&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/6746/a9fa1f89/ceb9c1ee9b6352421fe8645607810c42ab38fd49.svg)](https://app.codesee.io/r/reviews?pr=6746&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Commit:
refactor

Commit:
refactor

Commit:
refactor

Commit:
refactor

Commit:
refactor

Kilo59 commented on great_expectations/expectations/registry.py:
```diff
@@ -16,6 +16,10 @@
+from great_expectations.core.metric_function_types import (
+    MetricFunctionTypes,
+    MetricPartialFunctionTypes,
+)
```
These should be part of `if TYPE_CHECKING` imports

alexsherstinsky replied:
@Kilo59 Good catch -- moved (in next commit).

Kilo59 commented on great_expectations/self_check/util.py:
```diff
@@ -467,6 +467,24 @@
 # except ImportError:
 #     teradatasqlalchemy = None
 
+try:
+    from great_expectations.dataset import SqlAlchemyDataset
+    from great_expectations.dataset.sqlalchemy_dataset import SqlAlchemyBatchReference
+except ImportError:
+    SqlAlchemyDataset = None  # type: ignore[misc,assignment] # could be None
+    SqlAlchemyBatchReference = None  # type: ignore[misc,assignment] # could be None
+    logger.debug(
+        "Unable to load sqlalchemy dataset; install optional sqlalchemy dependency for support."
+    )
+
+try:
+    from great_expectations.dataset import SparkDFDataset
+except ImportError:
+    SparkDFDataset = None  # type: ignore[misc,assignment] # could be None
```
Why are we changing this to a try:except: import?

If we are doing type-checking on these they need to be available during the type-checking step.

alexsherstinsky replied:
@Kilo59 I believe that this is important for when the corresponding `Spark` and `SQLAlchemy` imports are not present (i.e., packages are not imported and hence unavailable).

NathanFarmer commented on great_expectations/core/metric_function_types.py:
```diff
@@ -0,0 +1,38 @@
+logger = logging.getLogger(__name__)
+
+
+class MetricFunctionTypes(enum.Enum):
```
Can we use the `str` mixin here like: `class MetricFunctionTypes(enum.Enum)` so we don't need to access `.value`?

alexsherstinsky replied:
@NathanFarmer I am not sure -- potentially, but it would be subject of a separate change, because this change is purely module refactoring.

NathanFarmer commented on great_expectations/core/metric_function_types.py:
```diff
@@ -0,0 +1,38 @@
+    AGGREGATE_VALUE = "value"  # "aggregate_value"
+
+
+class MetricPartialFunctionTypes(enum.Enum):
```
Same comment here

alexsherstinsky replied:
@NathanFarmer I am not sure -- potentially, but it would be subject of a separate change, because this change is purely module refactoring.  Please feel to try; thanks.

NathanFarmer commented on great_expectations/self_check/util.py:
```diff
@@ -467,6 +467,24 @@
 #     teradatasqlalchemy = None
 
+try:
+    from great_expectations.dataset import SqlAlchemyDataset
+    from great_expectations.dataset.sqlalchemy_dataset import SqlAlchemyBatchReference
```
Why do we need to try/except imports from our own library? Because SQL Alchemy might not be installed? Can we use [this pattern](https://github.com/great-expectations/great_expectations/pull/6731#discussion_r1064903709)?

alexsherstinsky replied:
@NathanFarmer Yes -- this is exactly for the case for when the corresponding Spark and SQLAlchemy imports are not present.  Is saying `from __future__ import __annotations__` at the top enough for solving this problem?

Kilo59 replied:
>is saying `rom __future__ import __annotations__` at the top enough for solving this problem?

No that will not solve it.
But why is it suddenly a problem if we were doing it before?
Does something break?

alexsherstinsky replied:
@Kilo59 Yes, in the branch for Multi-Batch, it breaks tests.  This is because the multi-batch implementation is less forgiving of sloppy code.  Hence, I had to create work-arounds for V2, which exercise the old, `Dataset` style code patterns in the `Validator`, which, correspondingly, impacts the testing parts of `self_check/util.py`.  Thanks.

Kilo59 commented on tests/integration/fixtures/split_and_sample_data/splitter_test_cases_and_fixtures.py:
```diff
@@ -6,6 +6,7 @@
 
 import pandas as pd
 
+from great_expectations.core.metric_domain_types import MetricDomainTypes
```
Why is this import being added without any other changes in the module?

alexsherstinsky replied:
@Kilo59 Deleted -- must be a typo (as it is not used).  Good catch, thanks.

Kilo59 commented:
I would feel better about these changes if we had better coverage of `metrics`

![image](https://user-images.githubusercontent.com/13108583/211578149-a04a27e4-33dd-4b8f-9fe3-4bce0b3bbdd9.png)

Kilo59 reviewed (changes requested):
Questions about imports.
Also, I'd like to have some type-coverage improvements.

alexsherstinsky commented:
> Questions about imports. Also, I'd like to have some type-coverage improvements.

@Kilo59 I addressed some of your comments.  As this is purely refactor of modules, if there is any additional coverage of metrics to be done, then it should be in a different pull request.  Thanks.

Commit:
cleanup

Commit:
Merge branch 'develop' into maintenance/GREAT-1358/GREAT-1527/alexsherstinsky/metrics/consolidate_metric_types_enums-2023_01_09-288

Kilo59 reviewed (approved):
I'd like to see the metrics submodule covered by flake8 before any more metrics changes go through.
But approving this PR.

Commit:
cleanpup

Commit:
Improving type checking for a metric.

github-actions[bot] committed:
Merge develop into maintenance/GREAT-1358/GREAT-1527/alexsherstinsky/metrics/consolidate_metric_types_enums-2023_01_09-288

github-actions[bot] committed:
Merge develop into maintenance/GREAT-1358/GREAT-1527/alexsherstinsky/metrics/consolidate_metric_types_enums-2023_01_09-288

github-actions[bot] committed:
Merge develop into maintenance/GREAT-1358/GREAT-1527/alexsherstinsky/metrics/consolidate_metric_types_enums-2023_01_09-288
